File: api/app/tanken/abrechnung.py
"""Abrechnung der E-Tankstelle — die Zusammenführung von Menge, Satz und
Zuordnung zu einer nutzerweisen Rechnung.

`_abrechnung` ist die eine Stelle, aus der Vorschau, Liste, PDF und Versand
schöpfen — damit alle dasselbe sagen."""
from __future__ import annotations

from calendar import monthrange
from datetime import date

from fastapi import HTTPException
from sqlmodel import Session

from .. import eauto
from ..models import Objekt
from .einstellungen import zuordnung_lesen
from .nutzer import nutzer_lesen, schluessel
from .perioden import abrechnungs_label, aktive_monate
from .posten import _posten_holen, buchungen, posten_als_buchungen
from .satz import EIGEN_RABATT, satz_ableiten
from .typen import Buchung
from .verlauf import verlauf, verlauf_summe
from .versand import _benzin_verbrauch


def abrechne(buchungen: list[Buchung], nutzer: list[dict],
             satz: float | None) -> list[dict]:
    """Je Nutzer: geladene Menge, Satz und Betrag.

    Der Satz ist abgeleitet (N148) und gilt für alle Ladungen des Zeitraums
    gleichermaßen — es gibt keinen Preis je Ladung mehr. Ist er ``None``,
    bleiben `betrag` und `satz` ebenfalls ``None``: die Mengen stehen da, aber
    kein Geldbetrag, den niemand belegen kann. **Eine 0,00 € wäre hier eine
    Behauptung**, keine Rechnung.

    Angelegte Nutzer ohne Ladung erscheinen mit 0 — sonst verschwände jemand
    aus der Liste, nur weil er ein Quartal lang nicht geladen hat. Ladungen auf
    Namen, die (noch) nicht in der Liste stehen, stehen am Ende mit
    ``angelegt: false``: übergehen wäre stilles Verschlucken von Geld."""
    zeilen: dict[str, dict] = {}
    for n in nutzer:
        zeilen[schluessel(n["name"])] = {
            "nutzer_id": n["id"], "name": n["name"], "email": n.get("email", ""),
            "angelegt": True, "anzahl": 0, "kwh": 0.0, "betrag": 0.0,
            "ladungen": []}

    for b in buchungen:
        s = schluessel(b.person)
        zeile = zeilen.get(s)
        if zeile is None:
            zeile = zeilen[s] = {
                "nutzer_id": None, "name": b.person or "—", "email": b.email,
                "angelegt": False, "anzahl": 0, "kwh": 0.0, "betrag": 0.0,
                "ladungen": []}
        zeile["anzahl"] += 1
        zeile["kwh"] += b.kwh
        if satz is not None:
            zeile["betrag"] += b.kwh * satz
        if not zeile["email"] and b.email:
            zeile["email"] = b.email
        zeile["ladungen"].append({
            "datum": b.tag.isoformat() if b.tag else None,
            "kwh": round(b.kwh, 3),
            "preis": None if satz is None else round(satz, 4),
            "betrag": None if satz is None else round(b.kwh * satz, 2)})

    fertig = []
    for zeile in zeilen.values():
        fertig.append({**zeile, "kwh": round(zeile["kwh"], 3),
                       "betrag": None if satz is None
                                 else round(zeile["betrag"], 2),
                       "satz": None if satz is None else round(satz, 4)})
    # Wer geladen hat, steht oben; danach alphabetisch — eine Reihenfolge, die
    # sich zwischen zwei Aufrufen nicht von selbst ändert.
    fertig.sort(key=lambda z: (-z["kwh"], schluessel(z["name"])))
    return fertig


def _abrechnung(session: Session, o: Objekt, slug: str, jahr: int,
                quartal: int, quartale: list[int] | None = None,
                aus: set[int] | None = None) -> dict:
    """Die Abrechnung eines Zeitraums — von Vorschau, Liste und Versand
    gemeinsam benutzt, damit alle drei dasselbe sagen.

    Abgerechnet wird ausschliesslich, was einer Person zugeordnet ist: die
    Wallbox weiss, wie viel geladen wurde, aber nicht von wem. Damit eine
    Abrechnung über 0 € nicht rätselhaft bleibt, steht daneben, wie viel im
    Zeitraum überhaupt geladen wurde (`geladen_kwh`) — die Lücke zwischen
    beiden Zahlen ist die Antwort auf „wieso passiert da nichts" (N143).

    Der Satz kommt aus der Stromkette (N148); die Mengen des Zeitraums sagen,
    wie stark Netz und eigener Strom darin gewichtet sind.

    `quartale` (mehrere zusammen) und `aus` (einzeln abgewählte Monate) sind der
    Quartals-Feinschliff aus N169. Ohne sie bleibt es beim einzelnen `quartal`
    — der bisherige Weg (auch der Autoversand ruft so)."""
    quartale = list(quartale) if quartale is not None else [quartal]
    aus = set(aus or ())
    try:
        monate = aktive_monate(quartale, aus)
    except ValueError as fehler:
        raise HTTPException(400, str(fehler)) from fehler
    if not monate:
        raise HTTPException(400, "Für die Abrechnung ist kein Monat ausgewählt "
                                 "— mindestens ein Monat muss bleiben.")
    aktiv = set(monate)
    von = date(jahr, monate[0], 1)
    bis = date(jahr, monate[-1], monthrange(jahr, monate[-1])[1])
    label = abrechnungs_label(jahr, quartale, aus)
    liste = nutzer_lesen(session, o)
    posten_roh, quelle, _ = _posten_holen(session, o, von, bis)
    # Ein aus dem Quartal abgewählter Monat fällt aus Menge und Betrag heraus
    # (N169) — nicht nur aus der Anzeige. Darum hier an der Quelle filtern.
    posten = [p for p in posten_roh
              if p.tag is None or p.tag.month in aktiv]
    zeilen_verlauf = [z for z in verlauf(posten, von, bis) if z["monat"] in aktiv]
    summe = verlauf_summe(zeilen_verlauf) if posten else {}
    satz = satz_ableiten(session, o.id, von, bis,
                         summe.get("extern_kwh"), summe.get("eigen_kwh"))
    # N165 — mit genau einem Nutzer gehören ihm automatisch alle Ladungen des
    # Zeitraums (aus den geladenen Mengen, nicht nur aus den namentlich
    # erfassten Sätzen). Bei mehreren bleibt es bei der Zuordnung über den Namen.
    auto_buch, automatisch = posten_als_buchungen(posten, liste)
    if automatisch:
        quelle_buch = auto_buch
    else:
        # Mehrere Nutzer: Zuordnung über den Zeitraum, wenn Regeln hinterlegt
        # sind; sonst wie bisher über den Namen. Ausgeschlossene Ladungen fallen
        # in beiden Fällen heraus (N165/2), abgewählte Monate ebenso (N169).
        regeln, ausschluss = zuordnung_lesen(session, o.slug, liste)
        quelle_buch = buchungen(session, o.id, von, bis, liste, regeln,
                                ausschluss, aktiv)
    zeilen = abrechne(quelle_buch, liste, satz.misch)
    if automatisch:
        for z in zeilen:
            z["automatisch"] = z["kwh"] > 0
    # N170 — je Nutzer aus geladenen kWh und seinem Verbrauch die gefahrenen km,
    # den Preis je 100 km und das energetische Benzin-Äquivalent. Ohne
    # hinterlegten Verbrauch bleiben die Größen None (keine erfundene Zahl).
    nach_id = {n["id"]: n for n in liste}
    for z in zeilen:
        n = nach_id.get(z.get("nutzer_id")) or {}
        modell = n.get("e_auto_modell", "")
        verbrauch = n.get("verbrauch_kwh_100km", 0.0)
        kz = eauto.kennzahlen(z["kwh"], z["betrag"], verbrauch or None)
        z["e_auto_modell"] = modell
        z["verbrauch_kwh_100km"] = verbrauch or 0.0
        z["km"], z["preis_100km"], z["liter_100km"] = (
            kz["km"], kz["preis_100km"], kz["liter_100km"])
        # N184b — die Empfänger-Anschrift, damit die Inline-Vorschau sie unter
        # dem Namen zeigen kann (wie das PDF). Eine noch nicht angelegte Person
        # (nutzer_id None) hat keine.
        z["strasse"] = n.get("strasse", "")
        z["plz"] = n.get("plz", "")
        z["ort"] = n.get("ort", "")
        # N184b — der Benzin-Kostenvergleich je Nutzer, exakt wie ihn das PDF
        # baut (`_pdf_und_name`): realer Verbrauch eines vergleichbaren Benziners
        # aus der KI (gecacht über `_benzin_verbrauch`) gegen den echten
        # E-Auto-Preis je 100 km. Ohne Verbrauch oder ohne belastbaren
        # Benzinwert bleibt er None — kein Vergleich, keine erfundene Zahl.
        z["benzin"] = (eauto.benzin_vergleich(
            _benzin_verbrauch(session, modell), kz["preis_100km"], kz["km"])
            if verbrauch and verbrauch > 0 else None)
    # N184b — die Objekt-Bankverbindung (der Betreiber, an den überwiesen wird);
    # None, wenn ungepflegt. Dieselbe Quelle wie das PDF (`_pdf_und_name`).
    konto = {"kontoinhaber": o.kontoinhaber or "", "iban": o.iban or "",
             "bank": o.bank or ""}
    zugeordnet = round(sum(z["kwh"] for z in zeilen), 3)
    return {"objekt": o.name, "jahr": jahr, "quartal": quartal,
            "quartale": sorted(set(quartale)), "aus_monate": sorted(aus),
            "monate_aktiv": monate,
            "label": label,
            "von": von.isoformat(), "bis": bis.isoformat(),
            "automatisch": automatisch,
            "satz": satz.misch, "satz_netz": satz.netz,
            "satz_eigen": satz.eigen, "satz_rabatt": EIGEN_RABATT,
            "satz_herkunft": satz.herkunft, "satz_grund": satz.grund,
            "nutzer": zeilen,
            "kwh_gesamt": zugeordnet,
            "betrag_gesamt": (None if satz.misch is None
                              else round(sum(z["betrag"] for z in zeilen), 2)),
            "geladen_kwh": summe.get("kwh"), "quelle": quelle,
            "offen_kwh": (None if not summe
                          else round(summe["kwh"] - zugeordnet, 2)),
            "eigen_prozent": summe.get("eigen_prozent"),
            # N170 — die Annahme des Benzin-Äquivalents, sichtbar mitgeführt.
            "benzin_kwh_pro_liter": eauto.BENZIN_KWH_PRO_LITER,
            # N184b — die Objekt-Bankverbindung für den Konto-Block der
            # Inline-Vorschau; None, wenn keine der drei Angaben gepflegt ist.
            "konto": (konto if any(konto.values()) else None)}


def _zeile_holen(daten: dict, nutzer_id: int, name: str) -> dict:
    """Die Abrechnungszeile eines Nutzers — über die Kennung oder den Namen."""
    for z in daten["nutzer"]:
        if (nutzer_id and z["nutzer_id"] == nutzer_id) or \
           (name and schluessel(z["name"]) == schluessel(name)):
            return z
    raise HTTPException(404, "Für diesen Nutzer gibt es keine Abrechnung.")

File: public/assets/tankstelle/abrechnung.js
/* abrechnung.js — die Abrechnungs-Karte: Kopf mit Automatik-Schalter (N193),
   der Ø-Satz-Block links (N148/N186/N195: Netz·kWh × Netzsatz = Netz·€,
   Eigen·kWh × Eigensatz = Eigen·€), die Zeitraum-Matrix rechts (aus matrix.js),
   die Nutzerliste mit Betragen + Sende-Buttons (N202: Zeitraum + Betrag im
   Knopf), Vorschau (aus blatt.js) und die Summe.

   Zusaetzlich orchestriert dieses Modul den Autoversand-Schalter, das Setzen
   und Loeschen der „abgerechnet"-Marker (N182) und laedt die Objekt-
   Einstellungen (Autoversand, Mehrnutzer-Regeln). Die Matrix-UI selbst und die
   abgerechnet-Marker-Helfer wohnen in matrix.js, weil sie logisch zur
   Zeitraum-Auswahl gehoeren. */
import { S, kwh, zahl, satzText, datum, zeigeFehler } from './state.js';
import { api, esc, eur, melde } from '../immo.js';
import { infoKnopfText, INFOS } from './infos.js';
import { zeitraumWahl, zeitraumParams, aktiveMonate, periodeKurz,
         nutzerPeriodeAbgerechnet, periodeAbgerechnet, jahrWert }
  from './matrix.js';
import { periodeMonate, periodeSumme, vorschauEntprellt } from './blatt.js';
import { ladungenZeigen } from './zuordnen.js';

/* N165/2 — der Autoversand-Schalter. Standardmaessig aus; eingeschaltet geht
   die Abrechnung einen Tag nach Quartalsende automatisch an jeden Nutzer mit
   Adresse, Ladung und Satz — als Mail mit PDF. Ein ruhiger Schalter, kein
   Warnton: es ist eine Erleichterung. */
export function abrechnungKopf() {
  // N193 — eine Kopfzeile statt Titel + separatem Schalter: „E-Tankstelle
  // Abrechnung" mit EINEM Info-i (die zwei früheren Erklärtexte zusammengeführt),
  // rechts der kompakte Automatik-Schalter. Das i steht neben dem <h3>, nicht im
  // <label> — sonst schaltete ein Klick darauf die Checkbox mit.
  const info = infoKnopfText('E-Tankstelle Abrechnung',
    INFOS.abrechnung.text + '\n\n' + INFOS.autoversand.text);
  return `<div class="abkopf">
    <span class="abkopf-t"><h3>E-Tankstelle Abrechnung</h3>${info}</span>
    <label class="switch">
      <span class="sw-label">Automatische Abrechnung</span>
      <input type="checkbox" data-autoversand ${S.einst.autoversand ? 'checked' : ''}>
      <span class="sw-track" aria-hidden="true"></span>
    </label>
  </div>`;
}

export async function einstellungenLaden() {
  try {
    const d = await api(`/tankstelle/${encodeURIComponent(S.objektSlug)}/einstellungen`);
    S.einst = { autoversand: !!d.autoversand, regeln: d.regeln || [],
                ausschluss: d.ausschluss || [] };
  } catch { S.einst = { autoversand: false, regeln: [], ausschluss: [] }; }
}

/* N182 — die abgerechnet-Marker des gewählten Jahres holen. Wird zu Beginn von
   `abrechnungZeigen` gerufen, damit Chips und Versand-Knöpfe zum Jahr passen. */
export async function abgerechnetLaden() {
  // N187 — ALLE Jahre holen (ohne `?jahr`), damit auch ein nicht gewähltes Jahr
  // seinen ✓ tragen kann. Ein Marker kann quartals- (monat==null) oder
  // monatsgenau (monat gesetzt) sein.
  try {
    const d = await api(`/tankstelle/${encodeURIComponent(S.objektSlug)}/abgerechnet`);
    S.abgerechnet = d.abgerechnet || [];
  } catch { S.abgerechnet = []; }
}

/* N182 — die gewählte Periode von Hand als abgerechnet setzen oder die Markierung
   entfernen (für den Initialstand der alten 4-Monats-Abrechnungen und für
   Korrekturen). Gilt für alle gelisteten Nutzer. */
export async function abgerechnetUmschalten() {
  if (!S.nutzerListe.length) {
    melde('Erst einen Nutzer anlegen', 'neg');
    return;
  }
  const monate = aktiveMonate();
  if (!monate.length) { melde('Kein Monat gewählt', 'neg'); return; }
  const setzen = !periodeAbgerechnet();
  try {
    // N193 — GENAU die wirksamen Monate markieren (nicht ganze Quartale): so
    // bleibt Nicht-Angewähltes unberührt. Das Backend setzt bei nicht-leerem
    // `monate` monatsgenaue Marker; `quartale` ist nur der Fallback (altes Backend).
    await api(`/tankstelle/${encodeURIComponent(S.objektSlug)}/abgerechnet`, {
      method: 'PUT',
      body: { jahr: jahrWert(), quartale: S.quartaleGewaehlt, monate,
              nutzer_id: 0, abgerechnet: setzen } });
    melde(setzen ? `${monate.length} Monat${monate.length === 1 ? '' : 'e'} als `
            + 'abgerechnet markiert' : 'Markierung entfernt', 'pos');
    await abrechnungZeigen();
  } catch (fehler) { melde(String(fehler.message || fehler), 'neg'); }
}

export async function autoversandUmschalten(aktiv) {
  try {
    await api(`/tankstelle/${encodeURIComponent(S.objektSlug)}/autoversand`,
              { method: 'PUT', body: { aktiv } });
    S.einst.autoversand = aktiv;
    melde(aktiv ? 'Automatischer Versand eingeschaltet'
                : 'Automatischer Versand ausgeschaltet', 'pos');
  } catch (fehler) {
    melde(String(fehler.message || fehler), 'neg');
    await abrechnungZeigen();               // Schalter zuruecksetzen
  }
}

/* Der Kern des Fehlers „ich druecke und nichts passiert": abgerechnet wird
   nur, was einer Person zugeordnet ist. Die Wallbox weiss, WIE VIEL geladen
   wurde, aber nicht von WEM. Ohne eine einzige Zuordnung zeigt jedes Quartal
   dieselben 0,00 € — und bisher stand nirgends, warum. Also: die Luecke
   zwischen geladen und zugeordnet benennen, statt sie stumm zu rendern. */
function offeneNotiz(d) {
  if (!(d.offen_kwh > 0.05)) return '';
  const nichts = !(d.kwh_gesamt > 0);
  return `<div class="notiz"><span class="wi">◔</span><span>Im Zeitraum
    ${esc(d.label)} wurden ${zahl(d.geladen_kwh)} kWh geladen, davon
    ${nichts ? 'keine einzige' : `nur ${zahl(d.kwh_gesamt)} kWh`} einer Person
    zugeordnet. Abgerechnet wird nur, was zugeordnet ist — die Wallbox weiß
    nicht, wer geladen hat. Das geht unten unter „Ladungen zuordnen“.</span></div>`;
}

/* N165 — ist nur eine Person angelegt, gehoeren ihr automatisch alle Ladungen
   des Zeitraums. Das ist eine Erleichterung, keine Stoerung: ein ruhiger
   Hinweis, kein Warnbalken — und die Zeile traegt zusaetzlich den Chip
   „automatisch", damit die Zuordnung nicht wie eine Handeingabe aussieht. */
function autoNotiz() {
  return `<div class="notiz auto"><span class="wi">⤳</span><span>Automatisch
    zugeordnet — nur eine Person angelegt, alle Ladungen gehen auf sie.</span></div>`;
}

/* Der Satz je kWh wird abgeleitet, nicht eingegeben (N148). An der Stelle, an
   der frueher ein Eingabefeld stand, steht jetzt die Zahl mit ihrer Herkunft:
   welcher Betrag, welche Menge, aus welcher Position welchen Zeitraums.
   Steht kein Satz fest, gibt es keine 0,00 € — sondern die Aussage, wodurch
   er entsteht. */
export function satzBlock(d) {
  // N187 — der Zeitraum, für den der Durchschnittssatz gilt, steht jetzt AM
  // Satz statt in einer eigenen Zeile darüber („dieser Ø-Wert gilt für …").
  const zr = `gilt für ${esc(d.label)} · ${datum(d.von)} – ${datum(d.bis)}`;
  if (d.satz == null) {
    return `<div class="notiz"><span class="wi">◔</span><span>${esc(
      d.satz_grund || 'Der Satz je kWh lässt sich noch nicht ermitteln.')}
      </span></div>
      <span class="satz-zr solo">${zr}</span>`;
  }
  // Jede Ergaenzung einzeln geprueft: fehlt eine, bleibt sie weg. Ein „NaN %"
  // oder ein „undefined" waere schlimmer als eine Zeile weniger.
  const rabatt = d.satz_rabatt == null ? ''
    : ` (${Math.round(d.satz_rabatt * 100)} % darunter)`;
  const hatTeile = d.satz_netz != null && d.satz_eigen != null;
  // N195 — statt nur der zwei Sätze zeigen die Karten die MENGEN und BETRÄGE des
  // gewählten Zeitraums: Netz-kWh × Netzsatz = Netz-€, Eigen-kWh × Eigensatz =
  // Eigen-€. So füllt der linke Block sinnvoll und die Herkunft des Mischsatzes
  // ist ablesbar. Fehlt die Aufteilung (unbekannt), bleibt die Menge „—".
  const s = periodeSumme(periodeMonate(d));
  const teil = (klasse, label, menge, satzc) => {
    const betrag = (menge != null && satzc != null) ? menge * satzc : null;
    return `<div class="satz-teil ${klasse}"><i></i>
      <div class="st-kopf"><span class="st-l">${label}</span>
        <span class="st-s">${satzText(satzc)}</span></div>
      <div class="st-rechnung"><span class="st-m">${menge == null ? '—' : kwh(menge)}</span>
        <span class="st-eq">→</span>
        <span class="st-v">${betrag == null ? '—' : eur(betrag)}</span></div></div>`;
  };
  const teile = !hatTeile ? '' : `<div class="satz-teile">
      ${teil('netz', 'Netz', s.extern_kwh, d.satz_netz)}
      ${teil('eigen', 'Eigen' + (rabatt ? ' · −10 %' : ''), s.eigen_kwh, d.satz_eigen)}
    </div>`;
  // Mischregel und Herkunft wandern ins i neben die Hero-Zahl (N186) — beide
  // bestehende Texte, nur an einem Ort statt als graue Textwand.
  const regel = !hatTeile ? ''
    : `Netz ${satzText(d.satz_netz)} · eigener Strom ${satzText(d.satz_eigen)}`
      + `${rabatt}. Der Satz einer Ladung mischt beide nach den Mengen des Zeitraums.`;
  const infoText = [regel, d.satz_herkunft || ''].filter(Boolean).join('\n\n');
  return `<div class="satz">
    <div class="satz-hero"><span class="sw">${satzText(d.satz)}</span>
      ${infoKnopfText('Satz je kWh', infoText)}</div>
    ${teile}
    <span class="satz-zr">${zr}</span></div>`;
}

export async function abrechnungZeigen() {
  const ziel = document.getElementById('abrechnung');
  if (!ziel) return;
  const vorher = S.offeneVorschau;         // eine offene Vorschau ueberlebt das Neuzeichnen
  let d;
  try {
    d = await api(`/tankstelle/${encodeURIComponent(S.objektSlug)}/abrechnung`
                  + `?${zeitraumParams()}`);
  } catch (fehler) {
    zeigeFehler(ziel, 'Abrechnung nicht verfügbar', fehler);
    return;
  }
  S.letzteAbrechnung = d;                  // die HTML-Vorschau liest daraus (N179)
  await abgerechnetLaden();                // N182 — Marker fürs gewählte Jahr

  const zeilen = (d.nutzer || []).map(n => `
    <div class="az" data-nutzer="${n.nutzer_id || 0}">
      <div class="ak">
        <span class="an">${esc(n.name)}${n.automatisch
          ? '<span class="chip teal" style="margin-left:8px">automatisch</span>' : ''}${
          n.angelegt ? ''
          : '<span class="chip amber" style="margin-left:8px">nicht angelegt</span>'}${
          nutzerPeriodeAbgerechnet(n.nutzer_id)
          ? '<span class="chip teal" style="margin-left:8px">✓ abgerechnet</span>' : ''}</span>
        <span class="ab">${n.betrag == null ? '—' : eur(n.betrag)}</span>
      </div>
      <span class="ad">${kwh(n.kwh)} · ${n.satz == null
          ? 'Satz steht noch nicht fest' : satzText(n.satz)}
        ${n.anzahl ? ` · ${n.anzahl} ${n.anzahl === 1 ? 'Ladung' : 'Ladungen'}` : ''}
        ${n.email ? ` · ${esc(n.email)}` : ''}</span>
      ${n.km != null ? `<div class="eco"><span class="ecar">🚗</span>
        <span class="ecochip"><span class="ev">${zahl(n.km, 0)}</span>
          <span class="eu">km</span></span>${
        n.preis_100km != null ? `<span class="ecochip">
          <span class="ev">${eur(n.preis_100km)}</span>
          <span class="eu">/ 100 km</span></span>` : ''}${
        n.liter_100km != null ? `<span class="ecochip">
            <span class="ev">≈ ${zahl(n.liter_100km, 1)} l</span>
            <span class="eu">/ 100 km</span></span>`
          + infoKnopfText('Benzin-Vergleich',
              `Annahme ${zahl(d.benzin_kwh_pro_liter, 1)} kWh je Liter · `
              + `reiner Energievergleich, kein Kostenvergleich`) : ''}</div>` : ''}
      ${n.email ? '' : `<div class="notiz fehlt"><span class="wi">✉</span>
        <span>Ohne E-Mail-Adresse lässt sich nichts verschicken.</span></div>`}
      ${n.kwh > 0 && n.nutzer_id ? `<div class="aw">
        <button class="btn leise" data-vorschau="${n.nutzer_id}">Vorschau</button>
        ${n.email && n.betrag != null
          ? (nutzerPeriodeAbgerechnet(n.nutzer_id)
            ? `<button class="btn gesperrt" disabled aria-disabled="true"
                 title="Diese Periode ist bereits abgerechnet">
                 <span class="bs-t">${periodeKurz()} · bereits abgerechnet</span>
                 <span class="bs-e">${eur(n.betrag)}</span></button>`
            : `<button class="btn btn-send" data-senden="${n.nutzer_id}">
                 <span class="bs-t">${periodeKurz()} · per Mail senden</span>
                 <span class="bs-e">${eur(n.betrag)}</span></button>`)
          : ''}
      </div>` : ''}
      <div data-mail="${n.nutzer_id || 0}"></div>
    </div>`).join('');

  ziel.innerHTML = `${abrechnungKopf()}
    <div class="absel">${zeitraumWahl()}${satzBlock(d)}</div>
    ${d.automatisch ? '' : offeneNotiz(d)}
    ${zeilen || `<div class="leer"><b>Noch nichts abzurechnen</b>
        Für ${esc(d.label)} gibt es weder Nutzer noch Ladungen.</div>`}
    ${d.automatisch && d.nutzer?.length ? autoNotiz() : ''}
    ${(d.nutzer?.length > 1 || d.offen_kwh > 0.05) ? `<div class="kpi" style="margin-top:14px">
      <div class="k"><span class="kl">${esc(d.label)}</span>
        <span class="kv">${kwh(d.kwh_gesamt)}</span></div>
      <div class="k"><span class="kl">Summe</span>
        <span class="kv teal">${d.betrag_gesamt == null ? '—'
          : eur(d.betrag_gesamt)}</span></div>
    </div>` : ''}`;
  // War eine Vorschau offen und gibt es den Nutzer noch, wird sie mit dem
  // neuen Ausschnitt neu geladen — entprellt, damit rasche Klicks (Monat an/aus)
  // nicht bei jedem Schritt ein PDF nachladen.
  S.offeneVorschau = 0;
  if (vorher && document.querySelector(`[data-mail="${vorher}"]`)) {
    S.offeneVorschau = vorher;
    vorschauEntprellt(vorher);
  }
  await ladungenZeigen();
}
